HB/structure.elpi

```diff
@@ -310,8 +310,41 @@ declare-coercion SortProjection ClassProjection
   log.coq.env.add-const-noimplicits SName SCoeBody STy @transparent! SC,
   log.coq.coercion.declare (coercion (const SC) 0 StructureF (grefclass StructureT)),
   log.coq.CS.declare-instance (const SC), % TODO: API in Elpi, take a @constant instead of gref
+
+  if-verbose (coq.say "HB: declare coercion path compression rules"),
+
+  findall-classes All,
+  CurrentTgtClass = (class TC StructureT TMLwP),
+  std.filter All (sub-class? CurrentTgtClass) AllTgtSuper,
+  std.map AllTgtSuper class_structure AllTgtSuperStructures,
+
+  mk-compression-clauses StructureF StructureT AllTgtSuperStructures AllCompressionClauses,
+  std.forall AllCompressionClauses (c\ log.coq.env.accumulate current "hb.db" (clause _ _ c)),
 ].
 
+
+pred mk-compression-clauses i:gref, i:gref, i:list gref, o:list prop.
+mk-compression-clauses StructureF StructureT [] [].
+mk-compression-clauses StructureF StructureT [StructureE|Rest] [Clause|Clauses]:- std.do! [
+  coq.coercion.db-for (grefclass StructureF) (grefclass StructureT) [pr C1 Nparams1],
+  coq.coercion.db-for (grefclass StructureT) (grefclass StructureE) [pr C2 Nparams2],
+  coq.coercion.db-for (grefclass StructureF) (grefclass StructureE) [pr C3 Nparams3],
```

Comment:
maybe this deserve an assertion, we do declare coercions in the right order, but if we break the invariant these commands may fail